llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -8308,6 +8367,115 @@ llvm::Function *SPIRVToLLVM::createLibraryEntryFunc() {
   return func;
 }
 
+const Vkgc::PipelineOptions *SPIRVToLLVM::getPipelineOptions() const {
+  assert(m_context && "Invalid context!");
+
+  return static_cast<Llpc::Context *>(m_context)->getPipelineContext()->getPipelineOptions();
+}
+
+template <> SPIRVValue *SPIRVToLLVM::getValueForScratchBoundaryCheckWrapper<SPIRVStore>(SPIRVStore *toBeWrapped) const {
+  return toBeWrapped->getDst();
+}
+
+template <> SPIRVValue *SPIRVToLLVM::getValueForScratchBoundaryCheckWrapper<SPIRVLoad>(SPIRVLoad *toBeWrapped) const {
+  return toBeWrapped->getSrc();
+}
+
+template <class T>
+std::pair<bool, SPIRVToLLVM::ScratchBoundaryCheckInfo> SPIRVToLLVM::prepareScratchBoundaryCheck(T *origin) {
+  auto const nullValue = std::make_pair<bool, ScratchBoundaryCheckInfo>(false, {});
+  if (!getPipelineOptions()->enableScratchAccessBoundaryChecks) {
+    return nullValue;
+  }
+
+  SPIRVValue *toBeWrapped = getValueForScratchBoundaryCheckWrapper(origin);
+  if (toBeWrapped == nullptr || toBeWrapped->getOpCode() != OpAccessChain) {
+    return nullValue;
+  }
+
+  SPIRVStorageClassKind pointerStorageClass = toBeWrapped->getType()->getPointerStorageClass();
+
+  if (pointerStorageClass != StorageClassPrivate && pointerStorageClass != StorageClassFunction) {
+    return nullValue;
+  }
+
+  auto const accessChainBase = static_cast<SPIRVAccessChainBase *const>(toBeWrapped);
+
+  std::vector<SPIRVValue *> indicesVec;
+
+  // Find any integer index, either for accessing the array or a struct member.
+  for (auto *const operand : accessChainBase->getOperands()) {
+    if (operand->getType()->isTypeInt()) {
+      indicesVec.push_back(operand);
+    }
+  }
+
+  if (!indicesVec.empty()) {
+    ArrayRef<SPIRVValue *> indices = indicesVec;
+    SPIRVValue *currentIndex = nullptr;
+
+    auto tryEatIndex = [&currentIndex, &indices]() -> void {
+      assert(!indices.empty());
+      currentIndex = indices[0];
+      indices = indices.slice(1);
+    };
+
+    auto baseType = accessChainBase->getBase()->getType()->getPointerElementType();
+    tryEatIndex();
+
+    Value *finalCmpResult = nullptr;
+
+    auto entryBlock = getBuilder()->GetInsertBlock();
+    Function *parent = entryBlock->getParent();
+    auto intermediateBlock = BasicBlock::Create(*m_context, "", parent);
+    auto finalBlock = BasicBlock::Create(*m_context, "", parent);
+
+    // Build a chain of OOB checks.
+    while (currentIndex) {
+      if (baseType->getOpCode() == OpTypeArray) {
+        if (auto const baseArray = static_cast<SPIRVTypeArray *const>(baseType)) {
```

Comment:
I think the convention is to put `const` before type name. I'm also not sure if `SPIRVTypeArray *const` does anything -- I think the const applies to the pointer itself, but `baseArray` is already `const`.

Reply:
The intention here is to have the `baseArray` immutable, so, I changed that, thanks for the hint. The pointer is already `const`, so having the modifier behind the pointer symbol does not add any value.